### ISSUE TITLE
Ambiguous Occurrence

### ISSUE THREAD

**Issue by sent1nel** [tomjaguarpaw/haskell-opaleye#394]
Doing a simple `cabal install opaleye` on OSX with GHC 8.4.3 yields an "ambiguous occurrence" error for `<>`

```
clang: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
Resolving dependencies...
Configuring opaleye-0.6.7000.0...
Building opaleye-0.6.7000.0...
Failed to install opaleye-0.6.7000.0
Build log ( ~/.cabal/logs/ghc-8.4.3/opaleye-0.6.7000.0-2zLcmx53TlnJY34QVAuO33.log ):
cabal: Entering directory '/var/folders/kv/4xxz5fvd7jq8pwkkr4j1dlj40000gn/T/cabal-tmp-46623/opaleye-0.6.7000.0'
Configuring opaleye-0.6.7000.0...
clang: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
Preprocessing library for opaleye-0.6.7000.0..
Building library for opaleye-0.6.7000.0..
[ 1 of 53] Compiling Opaleye.Internal.HaskellDB.Sql ( src/Opaleye/Internal/HaskellDB/Sql.hs, dist/build/Opaleye/Internal/HaskellDB/Sql.o )
[ 2 of 53] Compiling Opaleye.Internal.HaskellDB.Sql.Print ( src/Opaleye/Internal/HaskellDB/Sql/Print.hs, dist/build/Opaleye/Internal/HaskellDB/Sql/Print.o )

src/Opaleye/Internal/HaskellDB/Sql/Print.hs:123:39: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Prelude.<>’,
                             imported from ‘Prelude’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:7:8-43
                             (and originally defined in ‘GHC.Base’)
                          or ‘Text.PrettyPrint.HughesPJ.<>’,
                             imported from ‘Text.PrettyPrint.HughesPJ’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:30:53-56
    |
123 |     Just sn -> doubleQuotes (text sn) <> text "." <> tname
    |                                       ^^

src/Opaleye/Internal/HaskellDB/Sql/Print.hs:123:51: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Prelude.<>’,
                             imported from ‘Prelude’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:7:8-43
                             (and originally defined in ‘GHC.Base’)
                          or ‘Text.PrettyPrint.HughesPJ.<>’,
                             imported from ‘Text.PrettyPrint.HughesPJ’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:30:53-56
    |
123 |     Just sn -> doubleQuotes (text sn) <> text "." <> tname
    |                                                   ^^

src/Opaleye/Internal/HaskellDB/Sql/Print.hs:160:54: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Prelude.<>’,
                             imported from ‘Prelude’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:7:8-43
                             (and originally defined in ‘GHC.Base’)
                          or ‘Text.PrettyPrint.HughesPJ.<>’,
                             imported from ‘Text.PrettyPrint.HughesPJ’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:30:53-56
    |
160 |       CompositeSqlExpr s x   -> parens (ppSqlExpr s) <> text "." <> text x
    |                                                      ^^

src/Opaleye/Internal/HaskellDB/Sql/Print.hs:160:66: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Prelude.<>’,
                             imported from ‘Prelude’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:7:8-43
                             (and originally defined in ‘GHC.Base’)
                          or ‘Text.PrettyPrint.HughesPJ.<>’,
                             imported from ‘Text.PrettyPrint.HughesPJ’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:30:53-56
    |
160 |       CompositeSqlExpr s x   -> parens (ppSqlExpr s) <> text "." <> text x
    |                                                                  ^^

src/Opaleye/Internal/HaskellDB/Sql/Print.hs:162:46: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Prelude.<>’,
                             imported from ‘Prelude’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:7:8-43
                             (and originally defined in ‘GHC.Base’)
                          or ‘Text.PrettyPrint.HughesPJ.<>’,
                             imported from ‘Text.PrettyPrint.HughesPJ’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:30:53-56
    |
162 |       SubscriptSqlExpr e1 e2 -> ppSqlExpr e1 <> brackets (ppSqlExpr e2)
    |                                              ^^

src/Opaleye/Internal/HaskellDB/Sql/Print.hs:166:40: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Prelude.<>’,
                             imported from ‘Prelude’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:7:8-43
                             (and originally defined in ‘GHC.Base’)
                          or ‘Text.PrettyPrint.HughesPJ.<>’,
                             imported from ‘Text.PrettyPrint.HughesPJ’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:30:53-56
    |
166 |       FunSqlExpr f es        -> text f <> parens (commaH ppSqlExpr es)
    |                                        ^^

src/Opaleye/Internal/HaskellDB/Sql/Print.hs:171:45: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Prelude.<>’,
                             imported from ‘Prelude’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:7:8-43
                             (and originally defined in ‘GHC.Base’)
                          or ‘Text.PrettyPrint.HughesPJ.<>’,
                             imported from ‘Text.PrettyPrint.HughesPJ’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:30:53-56
    |
171 |       CastSqlExpr typ e      -> text "CAST" <> parens (ppSqlExpr e <+> text "AS" <+> text typ)
    |                                             ^^

src/Opaleye/Internal/HaskellDB/Sql/Print.hs:173:46: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Prelude.<>’,
                             imported from ‘Prelude’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:7:8-43
                             (and originally defined in ‘GHC.Base’)
                          or ‘Text.PrettyPrint.HughesPJ.<>’,
                             imported from ‘Text.PrettyPrint.HughesPJ’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:30:53-56
    |
173 |       ArraySqlExpr es        -> text "ARRAY" <> brackets (commaH ppSqlExpr es)
    |                                              ^^

src/Opaleye/Internal/HaskellDB/Sql/Print.hs:175:50: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Prelude.<>’,
                             imported from ‘Prelude’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:7:8-43
                             (and originally defined in ‘GHC.Base’)
                          or ‘Text.PrettyPrint.HughesPJ.<>’,
                             imported from ‘Text.PrettyPrint.HughesPJ’ at src/Opaleye/Internal/HaskellDB/Sql/Print.hs:30:53-56
    |
175 |       AggrFunSqlExpr f es ord distinct -> text f <> parens (ppSqlDistinct distinct <+> commaH ppSqlExpr es <+> ppOrderBy ord)
    |                                                  ^^
cabal: Leaving directory '/var/folders/kv/4xxz5fvd7jq8pwkkr4j1dlj40000gn/T/cabal-tmp-46623/opaleye-0.6.7000.0'
cabal: Error: some packages failed to install:
opaleye-0.6.7000.0-2zLcmx53TlnJY34QVAuO33 failed during the building phase.
The exception was:
ExitFailure 1
```

**Comment by sent1nel** [comment]
Interestingly, this is not an issue from my other laptop, a MacBook Pro running OSX Sierra (10.12.5).

**Comment by tomjaguarpaw** [comment]
It's fine in the test suite too https://travis-ci.org/tomjaguarpaw/haskell-opaleye/jobs/399080135#L705

I'd appreciate any digging down you can do on this.

**Comment by saurabhnanda** [comment]
Hit this one while upgrading to LTS-12.1 

I hit a similar issue in another package where it complained about `import Data.Monoid` being redundant. It seems like it has been added as a default export to Prelude.

**Comment by saurabhnanda** [comment]
Confirmed, while https://www.stackage.org/haddock/lts-12.1/base-4.11.1.0/Prelude.html#t:Semigroup contains `Semigroup.<>`, https://www.stackage.org/haddock/lts-11.17/base-4.10.1.0/Prelude.html doesn't


In the code above, which one did you want to use? Our build is stuck due to this bug and I can whip up a quick PR.

**Comment by saurabhnanda** [comment]
Looked at the code. The one in `Text.PrettyPrint.HughesPJ` is intended to be used. Although one should also investigate if this comment is still valid:

```
-- ---------------------------------------------------------------------------
-- Horizontal composition @<>@

-- We intentionally avoid Data.Monoid.(<>) here due to interactions of
-- Data.Monoid.(<>) and (<+>).  See
-- http://www.haskell.org/pipermail/libraries/2011-November/017066.html

-- | Beside.
-- '<>' is associative, with identity 'empty'.
(<>) :: Doc -> Doc -> Doc
(<>) = liftBinary (Ann.<>)
{-# INLINE (<>) #-}
```

**Comment by tomjaguarpaw** [comment]
If you want to import this operator qualified that would be fine by me.

**Comment by saurabhnanda** [comment]
And another related issue -- 

```

/Users/saurabhnanda/projects/haskell-opaleye/src/Opaleye/Internal/Table.hs:181:10: error:
    • No instance for (Semigroup (Zip a))
        arising from the superclasses of an instance declaration
    • In the instance declaration for ‘Monoid (Zip a)’
    |
181 | instance Monoid (Zip a) where
    |          ^^^^^^^^^^^^^^
```

**Comment by saurabhnanda** [comment]
@tomjaguarpaw please take a look at the PR

**Comment by tomjaguarpaw** [comment]
If anyone can explain why these issues don't show up in the test suite I would be grateful

https://travis-ci.org/tomjaguarpaw/haskell-opaleye/jobs/399080135#L705

**Comment by tomjaguarpaw** [comment]
OK, so it was a GHC 8.4 incompatibility.  I uploaded the following to fix it and deprecated 0.6.7000.0

https://hackage.haskell.org/package/opaleye-0.6.7001.0
